File: a2sv_contest_problems/F_Path_Prefixes.cpp
#include <string>
#include <vector>
#include <iostream>
#include <deque>
#include <tuple>
#include <algorithm>
#include <map>
#include <functional>

#define int long long

using namespace std;

int32_t main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int test_cases;
    cin >> test_cases;

    for (int test = 0; test < test_cases; ++test)
    {
        map<int, vector<tuple<int, int, int>>> graph;

        int n;
        cin >> n;

        for (auto i = 1; i < n; ++i)
        {
            int p, a, b;
            cin >> p >> a >> b;

            graph[p].push_back({i + 1, a, b});
        }

        vector<int> ans(n + 1, 0);
        vector<int> bi_sum = {0};

        auto solve = [&](int a_sum) -> int
        {
            auto best = 0;
            int l = 0, r = bi_sum.size() - 1;

            while (l <= r)
            {
                int mid = (l + r) / 2;
                auto val = bi_sum[mid];
                if (val <= a_sum)
                {
                    best = mid;
                    l = mid + 1;
                }
                else
                    r = mid - 1;
            }

            return best;
        };

        function<void(int, int)> dfs = [&](int node, int a)
        {
            ans[node] = solve(a);

            for (auto [nei, aj, bj] : graph[node])
            {
                bi_sum.push_back(bi_sum[bi_sum.size() - 1] + bj);
                dfs(nei, a + aj);
                bi_sum.pop_back();
            }
        };

        dfs(1, 0);
        for (auto i = 2; i < ans.size(); ++i)
        {
            cout << ans[i] << (i != ans.size() - 1 ? ' ' : '\n');
        }
    }
}

File: a2sv_contest_problems/F_Path_Prefixes.py
from collections import defaultdict
from sys import setrecursionlimit, stdin, stdout
import sys
import threading


def main():
    t = int(stdin.readline().strip())

    for _ in range(t):
        n = int(stdin.readline().strip())
        tree = defaultdict(list)

        for i in range(1, n):
            p, a, b = map(int, stdin.readline().split())
            tree[p].append((i + 1, a, b))

        ans = [0] * (n + 1)
        bi_sum = [0]

        def solve(a_sum):
            best = 0
            l, r = 0, len(bi_sum) - 1

            while l <= r:
                mid = (l + r) // 2
                val = bi_sum[mid]

                if val <= a_sum:
                    best = mid
                    l = mid + 1
                else:
                    r = mid - 1

            return best

        def dfs(node=1, a=0):
            ans[node] = solve(a)

            for (nei, aj, b) in tree[node]:
                bi_sum.append(bi_sum[-1] + b)
                dfs(nei, a + aj)
                bi_sum.pop()

        dfs()
        print(*ans[2:])


sys.setrecursionlimit(100 + (2 * 10 ** 5))
threading.stack_size(1 << 26)

main_thread = threading.Thread(target=main)
main_thread.start()
main_thread.join()
